### PR TITLE
F-Beta score Discussion Continuation and Pull Request

### PR DESCRIPTION
**Pull request by 2468cool**
Pull request for adding F-Beta score from issue #36 and also to continue the discussion from there.

From discussion: Talking about whether the output should be similar to Keras's F-Score metric or more in-line with Metrax's outputs such as its Precision and Recall. Keras gives an output equal to the batch size. Current Metrax implementation gives a single value regardless of batch size.

Failing Test cases:
- high_threshold_f16
- low_threshold_f16
- high_threshold_f32
- high_threshold_bf16
- low_threshold_bf16

### CODE REVIEW

File: src/metrax/classification_metrics.py [hunk 1]
```diff
@@ -575,3 +575,91 @@ def compute(self) -> jax.Array:
...
+    Attributes:
+        beta: The beta value used in the F-Score metric
+        precision: The precision value used in the F-Score metric
+        recall: The recall value used in the F-Score metric
+        precision_metric: Object that calculates the precision value
+        recall_metric: Object that calculates the recall value
+        threshold: The threshold value used in the F-Score metric
+    """
+
+    precision_metric: Precision.empty()
```
Comment (jshin1394) [inline src/metrax/classification_metrics.py:604]: should be just the type `Precision`.

File: src/metrax/classification_metrics.py [hunk 2]
```diff
@@ -575,3 +575,91 @@ def compute(self) -> jax.Array:
...
+    recall_metric: Recall.empty()
```
Comment (jshin1394) [inline src/metrax/classification_metrics.py:605]: Same, should be just the type `Recall`.

**Comment by 2468cool** [comment]
> Hi Benjamin, thank you so much for your PR.
> 
> Could you help me understand why some tests are passing but some aren't? Seems like the size of keras output and metrax output are both 1.

Keras's F-Score metric throws an error when not receiving a 2D input. I changed the way the y-true and y-pred are reshaped to be 2 dimensional arrays. The current setup has it so that everything seems like a batch size one input. Keras outputs an answer based on how large the batch size is. If it's given a batch size of 8, it'll output an array of 8 numbers. The current testing setup makes it only output one number. As for why those 5 test cases are currently failing is unknown to me. I'll keep looking into it.

This is also why I wanted to know how we want the Metrax F-Score to give it's output. Should it act like how Keras does and give a full array based on batch size or be like precision and recall and give a single number answer.

File: src/metrax/classification_metrics.py (hunk 1 above)
Comment (2468cool) [reply src/metrax/classification_metrics.py]: Changed in latest commit

File: src/metrax/classification_metrics.py [hunk 3]
```diff
@@ -575,3 +575,91 @@ def compute(self) -> jax.Array:
...
+    def empty(cls) -> 'FBetaScore':
+        return cls(
+            beta = 1.0,
+            threshold = 0.5,
+            precision_metric = Precision.empty(),
+            recall_metric = Recall.empty(),
+        )
+
+    @classmethod
+    def from_model_output(cls, predictions: jax.Array, labels: jax.Array,) -> 'FBetaScore':
```
[the remark this answers is not in the document]
Comment (2468cool) [reply src/metrax/classification_metrics.py]: Fixed in latest commit. Added ValueError checks as well to prevent incorrect values from being passed in.

**Comment by 2468cool** [comment]
Interestingly, while I was testing around with the test cases, setting all the failed test cases to have a beta of 1.0 causes them to pass.

**Comment by 2468cool** [comment]
> Interestingly, while I was testing around with the test cases, setting all the failed test cases to have a beta of 1.0 causes them to pass.
> 
> And I just noticed that _only_ the test cases that have a beta other than 1.0 are failing.
So after some further testing it appears that my implementation is not taking in the changed beta value and is instead using the default 1.0 value given. Changing this default makes it so that the other test cases with a changed beta value pass when the default is changed to match the test case. I'll have to figure out why it is not taking in the new beta value and keeping the original. We still should figure out how we want Metrax's F-Score output to be, single number or batch sized based.

**Comment by 2468cool** [comment]
> So after some further testing it appears that my implementation is not taking in the changed beta value and is instead using the default 1.0 value given. Changing this default makes it so that the other test cases with a changed beta value pass when the default is changed to match the test case. I'll have to figure out why it is not taking in the new beta value and keeping the original. We still should figure out how we want Metrax's F-Score output to be, single number or batch sized based.
>

So it appears that when using the test cases' "for logits, labels in zip(y_pred, y_true):" that the rest use is causing the issue. The beta is not being carried over after each merge. If I instead feed the y_true and y_pred directly into the from_model_output with the specified beta and threshold it works correctly and all test cases bass for the current implementation of the data reshaping.

**Comment by 2468cool** [comment]
New commit changes how FBetaScore calculates its answer to allow for outputs like Keras's version. All test cases pass for me without needing to modify the data to change output answer size.

**Comment by jshin1394** [comment]
Hi Benjamin! Could you please address and resolve the lastest round of comments? Thank you so much!

File: src/metrax/classification_metrics.py [hunk 4]
```diff
@@ -575,3 +575,142 @@ def compute(self) -> jax.Array:
     )
     # Threshold goes from 0 to 1, so trapezoid is negative.
     return jnp.trapezoid(tp_rate, fp_rate) * -1
+
+@flax.struct.dataclass
+class FBetaScore(clu_metrics.Metric):
+
```
Comment (jshin1394) [inline src/metrax/classification_metrics.py:581]: nit: remove extra space.

File: src/metrax/classification_metrics.py [hunk 5]
```diff
@@ -575,3 +575,142 @@ def compute(self) -> jax.Array:
...
+            false_positives = jnp.array(0, jnp.float32),
+            false_negatives = jnp.array(0, jnp.float32),
+        )
+
+    @classmethod
+    def from_model_output(
+            cls,
+            predictions: jax.Array,
+            labels: jax.Array,
+            beta = beta,
```
Comment (jshin1394) [inline src/metrax/classification_metrics.py:623]: should we set thie default to 1 as well? and add to the docstring that when only predictions and labels are given, it calculates F1 score.

File: src/metrax/classification_metrics.py [hunk 6]
```diff
@@ -575,3 +575,142 @@ def compute(self) -> jax.Array:
...
+            threshold = 0.5,) -> 'FBetaScore':
+        """Updates the metric.
+
+            Args:
+                threshold: threshold value to use in the F-Score metric a floating number.
```
Comment (jshin1394) [inline src/metrax/classification_metrics.py:628]: let's keep these in order in which they appear in from_model_output(pred, labels, beta, threshold)

File: src/metrax/classification_metrics.py [hunk 7]
```diff
@@ -575,3 +575,142 @@ def compute(self) -> jax.Array:
...
+                beta: beta value to use in the F-Score metric. A floating number.
+                predictions: A floating point 1D vector whose values are in the range [0,
+                  1]. The shape should be (batch_size,).
+                labels: True value. The value is expected to be 0 or 1. The shape should
+                  be (batch_size,).
+
+            Returns:
+                The true positives, false positives, and false negatives.
```
Comment (jshin1394) [inline src/metrax/classification_metrics.py:636]: The updated FBetaScore object.

File: src/metrax/classification_metrics.py [hunk 8]
```diff
@@ -575,3 +575,142 @@ def compute(self) -> jax.Array:
...
+    Computes the F-Beta score for the binary classification given 'predictions' and 'labels'.
+
+    Formula for F-Beta Score:
+        b2 = beta ** 2
+        f_beta_score = ((1 + b2) * (precision * recall)) / (precision * b2 + recall)
+
+    F-Beta turns into the F1 Score when beta = 1.0
+
+    Attributes:
+        beta: The beta value used in the F-Score metric
```
Comment (jshin1394) [inline src/metrax/classification_metrics.py:594]: keep these in order in which they appear below.

File: src/metrax/classification_metrics_test.py [hunk 9]
```diff
@@ -17,6 +17,7 @@
 import os
 os.environ['KERAS_BACKEND'] = 'jax'
 
+from src.metrax.classification_metrics import FBetaScore
```
Comment (jshin1394) [inline src/metrax/classification_metrics_test.py:20]: from metrax import FBetaScore

File: src/metrax/classification_metrics_test.py [hunk 10]
```diff
@@ -264,6 +274,41 @@ def test_aucroc(self, inputs, dtype):
         atol=1e-2 if y_true.dtype in (jnp.float16, jnp.bfloat16) else 1e-7,
     )
 
+# Testing function for F-Beta classification
+  # name, output true, output prediction, threshold, beta
+  @parameterized.named_parameters(
+      ('basic_f16', OUTPUT_LABELS, OUTPUT_PREDS_F16, 0.5, 1.0),
```
Comment (jshin1394) [inline src/metrax/classification_metrics_test.py:280]: can the test case name suffix indicate something about the beta value?

**Review by jshin1394 — COMMENTED** [review]
Thank you so much! :)

File: src/metrax/classification_metrics.py (hunk 5 above)
Comment (2468cool) [reply src/metrax/classification_metrics.py:623]: We could, but the beta the method variable is being set to is the class variable, which already defaults to 1.

**Comment by jshin1394** [comment]
Thank you so much!

**Comment by 2468cool** [comment]
You're welcome! If you need me to come back to this I'll be happy to help again!

**Comment by 2468cool** [comment]
This is a bit interesting. While running the individual test files works it seems running them together by calling pytest causes an issue. This seems to be different from what the error thrown during the CI / test phase of checks.

image

**Comment by jshin1394** [comment]
https://github.com/google/metrax/pull/106 should fix the dependency issue. Could you sync your branch and push again? Thank you so much! :)

**Comment by 2468cool** [comment]
Ok that should have synced and pushed.

**Comment by 2468cool** [comment]
Latest push fixes the ruff check on my end. It also adds a comment explaining why the merge function is commented out. TO summarize here though it's commented out since the 'from_model_output' function for FBetaScore should be able to take in the whole dataset at once instead of splitting it into chunks. I left it in, just commented out, incase we want to rework the function or need it down the road.

Let me know if anything else is needed and I will gladly help out :). It's been fun working on this and working with you!
